ha/core/controllers/pcs/node_controller.py

```diff
@@ -242,12 +304,45 @@ def start(self, nodeid: str) -> dict:
             Log.error(f"{nodeid} status is {_node_status}, node may not be started.")
             raise ClusterManagerError(f"Failed to start {nodeid} as found unhandled status {_node_status}")
 
-class PcsHWNodeController(PcsVMNodeController):
+    @controller_error_handler
+    def stop(self, node_id: str, timeout: int= -1, **op_kwargs) -> dict:
+        """
+        Stop Node with nodeid.
+        Args:
+            nodeid (str): Node ID from cluster nodes.
+        Returns:
+            ([dict]): Return dictionary. {"status": "", "output": "", "error": ""}
+                status: Succeeded, Failed, InProgress
+        """
+        timeout = const.NODE_STOP_TIMEOUT if timeout < 0 else timeout
+        node_status = self.nodes_status([node_id]).get(node_id)
+        if node_status == NODE_STATUSES.CLUSTER_OFFLINE.value:
+            Log.info(f"For stop {node_id}, Node already in offline state.")
+            status = f"Node {node_id} is already in offline state."
+        elif node_status == NODE_STATUSES.POWEROFF.value:
+            raise ClusterManagerError(f"Failed to stop {node_id}."
+                f"node is in {node_status}.")
+        else:
+            if self.heal_resource(node_id):
+                time.sleep(const.BASE_WAIT_TIME)
+            try:
+                Log.info(f"Please Wait, trying to stop node: {node_id}")
+                self._execute.run_cmd(const.PCS_STOP_NODE.replace("<node>", node_id)
+                        .replace("<seconds>", str(timeout)))
+                Log.info(f"Executed node stop for {node_id}, Waiting to stop resource")
+                time.sleep(const.BASE_WAIT_TIME)
+                status = f"Stop for {node_id} is in progress, waiting to stop resource"
+            except Exception as e:
+                raise ClusterManagerError(f"Failed to stop {node_id}, Error: {e}")
+        return {"status": const.STATUSES.IN_PROGRESS.value, "output": status, "error": ""}
+
+class PcsHWNodeController(PcsNodeController):
     def initialize(self, controllers):
         """
         Initialize the storageset controller
         """
         self._controllers = controllers
+        self.ipmifenceagent = IpmiFencingAgent()
```

Comment:
self.fencing_agent= IpmiFencingAgent()

PcsHWNodeController is not aware of ipmi

Reply:
Yes, Changed accordingly